### ISSUE TITLE
np.fft.fft2 results are different on numpy v1.12.1 and v1.16.0 in the order e-10 

### ISSUE THREAD

**Issue by ddoe777** [numpy/numpy#14409]
I am seeing a difference in results for np.fft.fft2 in the order of 1e-11 on both real and imaginary values running it on two versions of numpy- v1.12.1 and v1.16.0. According to my test, similar thing happens for v1.14.0 (More consistent with v1.16.0)

The results of image_fft on v1.12.1 and v1.16 are different when compared. One way I compared was taking a sum of the entire image_fft. 
For v1.12.1 I found it to be: (60000.00000000001-7.275957614183426e-12j)
For v1.16.0 I found it to be: (59999.99999999999+2.1827872842550278e-11j)

The environment in which I am producing the results are exactly same and are as follows:
- Docker container of Ubuntu 16.04 LTS
- Python details are: 
Python 3.5.2 |Anaconda custom (64-bit)| (default, Jul  2 2016, 17:53:06)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux

I had v1.12.1 installed first on numpy and next I installed numpy == 1.16.0 using pip install numpy==1.16.0

Let me know more infos are required.
### Reproducing code example:

A way to generate it is:
```python
import numpy as np
image = np.ones((150,100))
(a,b) = image.shape
image_tmp = np.zeros(shape=(2*a,2*b))
image_tmp[0:a,0:b]=image
image_fft = np.fft.fft2(image_tmp)
```

### Numpy/Python version information:

**For v1.12.1:**
1.12.1 3.5.2 |Anaconda custom (64-bit)| (default, Jul  2 2016, 17:53:06)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]

**For v1.16.0:**
1.16.0 3.5.2 |Anaconda custom (64-bit)| (default, Jul  2 2016, 17:53:06)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]

**Comment by ddoe777** [comment]
[More details]
I checked through different versions of numpy starting from v1.12.1, it looks like at v1.14.0, the sum output looked different. However, that was due to the changes on array printing; which can be changed by doing np.set_printoptions(legacy="1.13"). However, the output matches through v1.12.1 to v1.14.5. 

The output stops being equal from v1.15.0rc1 onwards (https://github.com/numpy/numpy/tags?after=v1.16.0). I couldn't find anything in the doc (at first glance) that is related to this: https://github.com/numpy/numpy/releases/tag/v1.15.0rc1. This may be written in disguise in the notes. Can someone take a look at this?

**Comment by seberg** [comment]
Hmmm, I think that 1.15.x are the first versions with Intel patches in Anaconda, if you use `mkl_fft` or maybe `mkl` in general. NumPy itself should change/imrpove with 1.17 not 1.15 probably.

However, the differences you are reporting seem to be well within expected numerical stability. So to be honest, I do not think you can expect identical results (also for example on different hardware).

**Comment by ddoe777** [comment]
Thanks for the info on MKL. It's true that the results are within an expected numerical stability; but in my case, the algorithm is such that the error accumulates and I have seen it to spike to difference of order e-7. The unfortunate part about precision is that it's not a common constant that is different from value to value. and you are right, if I try it on different hardware will lead me to not identical result. However, I have my clients expecting to match the output on a single machine type and architecture (at time, it can get crazy!)

I would have been good using v1.12 for my operations; however I need to use tensorflow v1.14 which requires numpy v1.16.0. Unfortunately, multiple versions of a module are not allowed to be used in python. And anything of a method that allows to do that uses hack which includes changing library code too. I guess at this point, my best bet is to install numpy v1.16 from source. 

**Few questions:**
1. Is there a way to disable Intel MKL (even though it improves performance!) on numpy? 
2. Is there a detailed document on what is considered numerical stability for numpy?
3. Does numpy test if outputs from version to version change for the same operation due to a change in library (within a range of precision)?

**Comment by seberg** [comment]
Yes, you can uninstall `intel_mkl` or `mkl_fft` (or both) in conda I think. Another point maybe, 1.17 has much faster (typically) and numerically higher precision FFT. I guess if the thing is that multiple runs give different results, than I could imagine MKL exacerbates such effects. Still, it seem strange to me that you cannot check for almost equality, but that is something you have to know.

**Comment by ddoe777** [comment]
Thanks again for the info. Just to make it clear, multiple runs on the same input pipeline doesn't give different results. I meant that the difference in my case of algorithm which takes multiple images (unlike just one example as above- I created the above example to show the difference) and the difference of the output from the algorithm accumulates to a higher order. The point is for me to know what are the changes made from v1.12 to v1.16 and how I can address that (I guess StackOverFlow is a better place for the later since this is more for issues). 

I am looking at if no-mkl or disabling mkl_fft makes any difference. Will be back with update.

**Comment by ddoe777** [comment]
I did a little more digging. Few things I did:
- I tried uninstalling mkl-service from conda using conda remove mkl mkl-service and reinstalling numpy 16.0.0. However that didn't help. 
- Looking a little further into the configuration numpy (using import numpy as np; np.__config__.show()), I didn't find anything related to MKL: `blas_mkl_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
    library_dirs = ['/usr/local/lib']
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
    library_dirs = ['/usr/local/lib']
blis_info:
  NOT AVAILABLE
blas_opt_info:
    libraries = ['openblas', 'openblas']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
    library_dirs = ['/usr/local/lib']
lapack_mkl_info:
  NOT AVAILABLE
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
    library_dirs = ['/usr/local/lib']` 

Funnily enough, on my numpy version of v1.12.0 there was SCIPY_MKL_H: `lapack_opt_info:
    include_dirs = ['/home/soft/anaconda/include']
    library_dirs = ['/home/soft/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    libraries = ['mkl_rt', 'pthread']
blas_mkl_info:
    include_dirs = ['/home/soft/anaconda/include']
    library_dirs = ['/home/soft/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    libraries = ['mkl_rt', 'pthread']
lapack_mkl_info:
    include_dirs = ['/home/soft/anaconda/include']
    library_dirs = ['/home/soft/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    libraries = ['mkl_rt', 'pthread']
blas_opt_info:
    include_dirs = ['/home/soft/anaconda/include']
    library_dirs = ['/home/soft/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    libraries = ['mkl_rt', 'pthread']`

**Finally,** I dug into the code of numpy's fft and found that there is fftpack.py  being used for fft calls which uses fftpack.c at the end. In the build version of numpy (both v1.12.1 and v1.16.0), there is a fftpack_lite.cpython-35m-x86_64-linux-gnu.so file. I swapped the .so file of v1.16.0 with v1.12.1 to see if there any change. Luckily there was no missing for running the above numpy code and the output turned out to be equal to the one in v1.12.1. 

**Digging deeper into the code** (https://github.com/numpy/numpy/compare/maintenance%2F1.12.x...maintenance%2F1.16.x?diff=split#diff-8f6f655f53f5622f2837e7fdfc83c7d7R93), I found that it's not Intel MKL FFT related issue. It's mainly subtle change of value for TWOPI - v1.12.1 has it locally called as = 6.28318530717959; whereas v1.16.0 has it macro-ed at 6.283185307179586476925286766559005768391 and uses cos2pi/sin2pi depending on the call. There are few other value changes with more precision as can be seen from the comparison above. Most changes are in the file: numpy/fft/fftpack.c. 

**Rough Summary:** Changes in value are not due to Intel MKL FFT; but changes in values of fftpack used by numpy itself. I haven't tested it on C-level; however the process should explain it enough. Let me know if I am missing anything.

**Comment by seberg** [comment]
Oh, 1.16.x! I thought the change was in 1.15.x. Yes, there were some fixes in the twiddle factors that I think went in before the larger changes/updates to FFT in 1.17 (not sure if that actually hit the release notes, but I would think so). Those changes are fixes (i.e. the old values are less precise than the new ones). So it should be preferable to use the new values.

**Comment by ddoe777** [comment]
The same goes for 1.15.x (as seen from here: https://github.com/numpy/numpy/compare/maintenance/1.12.x...maintenance/1.15.x). Yes you are right! They look like fixes for the twiddle factor and more precise. However, I didn't find those changes being mentioned in the release notes. IMO, this sort of minor fixes should be mentioned.

Regarding 1.17.x, it looks like there is a huge change on FFT (https://github.com/numpy/numpy/compare/maintenance/1.12.x...maintenance/1.17.x) and it looks like a simple replacement (possible between v1.12.x and v1.17.x) may not work if someone wants to match the output to a previous version.

**Comment by seberg** [comment]
I am surprised it did not make it into the release notes, although it was just bug fix in a sense. But in any case, I do not think there is much to be done here. These tiny changes can happen, and in this case they were even a fix, though.
